In [224]:
import pandas as pd

## Entrada dos dados

In [225]:
results = pd.read_csv('results.csv')

**Visualização dos dados**

In [226]:
results

date home_team    away_team  home_score  away_score  \
0      1872-11-30  Scotland      England           0           0   
1      1873-03-08   England     Scotland           4           2   
2      1874-03-07  Scotland      England           2           1   
3      1875-03-06   England     Scotland           2           2   
4      1876-03-04  Scotland      England           3           0   
...           ...       ...          ...         ...         ...   
44336  2022-12-04   England      Senegal           3           0   
44337  2022-12-05     Japan      Croatia           1           1   
44338  2022-12-05    Brazil  South Korea           4           1   
44339  2022-12-06   Morocco        Spain           0           0   
44340  2022-12-06  Portugal  Switzerland           6           1   

           tournament       city   country  neutral  
0            Friendly    Glasgow  Scotland    False  
1            Friendly     London   England    False  
2            Friendly    Glasgow  Scotland    False  
3            Friendly     London   England    False  
4            Friendly    Glasgow  Scotland    False  
...               ...        ...       ...      ...  
44336  FIFA World Cup    Al Khor     Qatar     True  
44337  FIFA World Cup  Al Wakrah     Qatar     True  
44338  FIFA World Cup       Doha     Qatar     True  
44339  FIFA World Cup  Al Rayyan     Qatar     True  
44340  FIFA World Cup     Lusail     Qatar     True  

[44341 rows x 9 columns]

In [227]:
results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44341 entries, 0 to 44340
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   date        44341 non-null  object
 1   home_team   44341 non-null  object
 2   away_team   44341 non-null  object
 3   home_score  44341 non-null  int64 
 4   away_score  44341 non-null  int64 
 5   tournament  44341 non-null  object
 6   city        44341 non-null  object
 7   country     44341 non-null  object
 8   neutral     44341 non-null  bool  
dtypes: bool(1), int64(2), object(6)
memory usage: 2.7+ MB


- Fazendo uma análise dos dados, podemos ver que a coluna 'date' está como object. Sendo assim, precisamos,na hora da importarção dos dados, o type do dado com o atributo parse_dates.

In [228]:
results = pd.read_csv('results.csv',parse_dates=['date'])

In [229]:
results

date home_team    away_team  home_score  away_score  \
0     1872-11-30  Scotland      England           0           0   
1     1873-03-08   England     Scotland           4           2   
2     1874-03-07  Scotland      England           2           1   
3     1875-03-06   England     Scotland           2           2   
4     1876-03-04  Scotland      England           3           0   
...          ...       ...          ...         ...         ...   
44336 2022-12-04   England      Senegal           3           0   
44337 2022-12-05     Japan      Croatia           1           1   
44338 2022-12-05    Brazil  South Korea           4           1   
44339 2022-12-06   Morocco        Spain           0           0   
44340 2022-12-06  Portugal  Switzerland           6           1   

           tournament       city   country  neutral  
0            Friendly    Glasgow  Scotland    False  
1            Friendly     London   England    False  
2            Friendly    Glasgow  Scotland    False  
3            Friendly     London   England    False  
4            Friendly    Glasgow  Scotland    False  
...               ...        ...       ...      ...  
44336  FIFA World Cup    Al Khor     Qatar     True  
44337  FIFA World Cup  Al Wakrah     Qatar     True  
44338  FIFA World Cup       Doha     Qatar     True  
44339  FIFA World Cup  Al Rayyan     Qatar     True  
44340  FIFA World Cup     Lusail     Qatar     True  

[44341 rows x 9 columns]

In [230]:
results.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 44341 entries, 0 to 44340
Data columns (total 9 columns):
 #   Column      Non-Null Count  Dtype         
---  ------      --------------  -----         
 0   date        44341 non-null  datetime64[ns]
 1   home_team   44341 non-null  object        
 2   away_team   44341 non-null  object        
 3   home_score  44341 non-null  int64         
 4   away_score  44341 non-null  int64         
 5   tournament  44341 non-null  object        
 6   city        44341 non-null  object        
 7   country     44341 non-null  object        
 8   neutral     44341 non-null  bool          
dtypes: bool(1), datetime64[ns](1), int64(2), object(5)
memory usage: 2.7+ MB


Analise de quantos torneios temos nesta planilha.

In [231]:
results['tournament'].nunique()

140

- Analisamos que tem 140 torneios, então devemos filtrar somente os dados da Copa do Mundo que, neste caso, são os que nos interessam

In [232]:
results['tournament'].unique()

array(['Friendly', 'British Home Championship', 'Évence Coppée Trophy',
       'Copa Lipton', 'Copa Newton', 'Copa Premio Honor Argentino',
       'Copa Premio Honor Uruguayo', 'Copa Roca', 'Copa América',
       'Copa Chevallier Boutell', 'Nordic Championship',
       'Central European International Cup', 'Baltic Cup', 'Balkan Cup',
       'FIFA World Cup', 'Copa Rio Branco',
       'FIFA World Cup qualification', 'Bolivarian Games',
       'CCCF Championship', 'NAFU Championship', 'Copa Oswaldo Cruz',
       'Pan American Championship', 'Copa del Pacífico',
       "Copa Bernardo O'Higgins", 'AFC Asian Cup qualification',
       'Atlantic Cup', 'AFC Asian Cup', 'African Cup of Nations',
       'Copa Paz del Chaco', 'Merdeka Tournament',
       'UEFA Euro qualification', 'UEFA Euro',
       'Windward Islands Tournament',
       'African Cup of Nations qualification', 'Vietnam Independence Cup',
       'Copa Carlos Dittborn', 'Phillip Seaga Cup',
       'CONCACAF Championship', 'Copa Ju

In [233]:
pd.DataFrame(results['tournament'].unique())

0
0                             Friendly
1            British Home Championship
2                 Évence Coppée Trophy
3                          Copa Lipton
4                          Copa Newton
..                                 ...
135              Mahinda Rajapaksa Cup
136                         Navruz Cup
137         CONIFA Africa Football Cup
138  CONIFA South America Football Cup
139           MSG Prime Minister's Cup

[140 rows x 1 columns]

- Salvando os dados da Copa do Mundo em uma nova variável.

In [234]:
results_fifa = results[results['tournament']=='FIFA World Cup']

In [235]:
results_fifa

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    Belgium  United States           0           3   
1312  1930-07-13     France         Mexico           4           1   
1313  1930-07-14     Brazil     Yugoslavia           1           2   
1314  1930-07-14       Peru        Romania           1           3   
1315  1930-07-15  Argentina         France           1           0   
...          ...        ...            ...         ...         ...   
44336 2022-12-04    England        Senegal           3           0   
44337 2022-12-05      Japan        Croatia           1           1   
44338 2022-12-05     Brazil    South Korea           4           1   
44339 2022-12-06    Morocco          Spain           0           0   
44340 2022-12-06   Portugal    Switzerland           6           1   

           tournament        city  country  neutral  
1311   FIFA World Cup  Montevideo  Uruguay     True  
1312   FIFA World Cup  Montevideo  Uruguay     True  
1313   FIFA World Cup  Montevideo  Uruguay     True  
1314   FIFA World Cup  Montevideo  Uruguay     True  
1315   FIFA World Cup  Montevideo  Uruguay     True  
...               ...         ...      ...      ...  
44336  FIFA World Cup     Al Khor    Qatar     True  
44337  FIFA World Cup   Al Wakrah    Qatar     True  
44338  FIFA World Cup        Doha    Qatar     True  
44339  FIFA World Cup   Al Rayyan    Qatar     True  
44340  FIFA World Cup      Lusail    Qatar     True  

[956 rows x 9 columns]

**Agora vamos criar uma nova coluna chaman Year, para pegar somente o ano em que a copa aconteceu, pois nós só precisamos do ano para fazer nossa análise**

- Analisando os anos

In [236]:
results_fifa['date'].dt.year

1311     1930
1312     1930
1313     1930
1314     1930
1315     1930
         ... 
44336    2022
44337    2022
44338    2022
44339    2022
44340    2022
Name: date, Length: 956, dtype: int64

- Criando coluna

In [237]:
results_fifa['year'] = results_fifa['date'].dt.year

C:\Users\vande\AppData\Local\Temp\ipykernel_9484\451678527.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  results_fifa['year'] = results_fifa['date'].dt.year


In [238]:
results_fifa

date  home_team      away_team  home_score  away_score  \
1311  1930-07-13    Belgium  United States           0           3   
1312  1930-07-13     France         Mexico           4           1   
1313  1930-07-14     Brazil     Yugoslavia           1           2   
1314  1930-07-14       Peru        Romania           1           3   
1315  1930-07-15  Argentina         France           1           0   
...          ...        ...            ...         ...         ...   
44336 2022-12-04    England        Senegal           3           0   
44337 2022-12-05      Japan        Croatia           1           1   
44338 2022-12-05     Brazil    South Korea           4           1   
44339 2022-12-06    Morocco          Spain           0           0   
44340 2022-12-06   Portugal    Switzerland           6           1   

           tournament        city  country  neutral  year  
1311   FIFA World Cup  Montevideo  Uruguay     True  1930  
1312   FIFA World Cup  Montevideo  Uruguay     True  1930  
1313   FIFA World Cup  Montevideo  Uruguay     True  1930  
1314   FIFA World Cup  Montevideo  Uruguay     True  1930  
1315   FIFA World Cup  Montevideo  Uruguay     True  1930  
...               ...         ...      ...      ...   ...  
44336  FIFA World Cup     Al Khor    Qatar     True  2022  
44337  FIFA World Cup   Al Wakrah    Qatar     True  2022  
44338  FIFA World Cup        Doha    Qatar     True  2022  
44339  FIFA World Cup   Al Rayyan    Qatar     True  2022  
44340  FIFA World Cup      Lusail    Qatar     True  2022  

[956 rows x 10 columns]

### Separação dos paises e gols.
Nós precisamos dos paises, quantidade de gols e ano.
Como está dividido em home_team e away_team, precisamos acabar com essa separação, pois será necessários para somarmos a quantidade de gols total que um país fez.

País da casa

In [239]:
home_score = results_fifa[['home_team','home_score','year']]
home_score

home_team  home_score  year
1311     Belgium           0  1930
1312      France           4  1930
1313      Brazil           1  1930
1314        Peru           1  1930
1315   Argentina           1  1930
...          ...         ...   ...
44336    England           3  2022
44337      Japan           1  2022
44338     Brazil           4  2022
44339    Morocco           0  2022
44340   Portugal           6  2022

[956 rows x 3 columns]

País visitante

In [240]:
away_score = results_fifa[['away_team','away_score','year']]
away_score

away_team  away_score  year
1311   United States           3  1930
1312          Mexico           1  1930
1313      Yugoslavia           2  1930
1314         Romania           3  1930
1315          France           0  1930
...              ...         ...   ...
44336        Senegal           0  2022
44337        Croatia           1  2022
44338    South Korea           1  2022
44339          Spain           0  2022
44340    Switzerland           1  2022

[956 rows x 3 columns]

- Para juntarmos as duas tabelas precisamos modificar o nome das colunas e deixá-los iguais.\
Para isso usamos o 'rename'.

País da casa

In [241]:
home_score = results_fifa[['home_team','home_score','year']].rename({'home_team':'team','home_score':'score'},axis=1)
home_score

team  score  year
1311     Belgium      0  1930
1312      France      4  1930
1313      Brazil      1  1930
1314        Peru      1  1930
1315   Argentina      1  1930
...          ...    ...   ...
44336    England      3  2022
44337      Japan      1  2022
44338     Brazil      4  2022
44339    Morocco      0  2022
44340   Portugal      6  2022

[956 rows x 3 columns]

País visitante

In [242]:
away_score = results_fifa[['away_team','away_score','year']].rename({'away_team':'team','away_score':'score'},axis=1)
away_score

team  score  year
1311   United States      3  1930
1312          Mexico      1  1930
1313      Yugoslavia      2  1930
1314         Romania      3  1930
1315          France      0  1930
...              ...    ...   ...
44336        Senegal      0  2022
44337        Croatia      1  2022
44338    South Korea      1  2022
44339          Spain      0  2022
44340    Switzerland      1  2022

[956 rows x 3 columns]

- Agora que as planilhas estão com nomes iguais, podemos uni-las com o 'concat'.

In [243]:
score_fifa = pd.concat([home_score,away_score])
score_fifa

team  score  year
1311       Belgium      0  1930
1312        France      4  1930
1313        Brazil      1  1930
1314          Peru      1  1930
1315     Argentina      1  1930
...            ...    ...   ...
44336      Senegal      0  2022
44337      Croatia      1  2022
44338  South Korea      1  2022
44339        Spain      0  2022
44340  Switzerland      1  2022

[1912 rows x 3 columns]

- Ajustando o index

In [244]:
score_fifa.reset_index(drop=True,inplace=True)

In [245]:
score_fifa

team  score  year
0         Belgium      0  1930
1          France      4  1930
2          Brazil      1  1930
3            Peru      1  1930
4       Argentina      1  1930
...           ...    ...   ...
1907      Senegal      0  2022
1908      Croatia      1  2022
1909  South Korea      1  2022
1910        Spain      0  2022
1911  Switzerland      1  2022

[1912 rows x 3 columns]

- Analisando de existem valores nulos na planilha

In [246]:
score_fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1912 entries, 0 to 1911
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   team    1912 non-null   object
 1   score   1912 non-null   int64 
 2   year    1912 non-null   int64 
dtypes: int64(2), object(1)
memory usage: 44.9+ KB


In [247]:
score_fifa.isnull().sum()

team     0
score    0
year     0
dtype: int64

In [248]:
score_fifa.isna().sum()

team     0
score    0
year     0
dtype: int64

- Verificando quantos times tem na planilha.

In [249]:
score_fifa.team.nunique()

82

In [250]:
score_fifa.team.unique()

array(['Belgium', 'France', 'Brazil', 'Peru', 'Argentina', 'Chile',
       'Bolivia', 'Paraguay', 'Uruguay', 'Austria', 'Czechoslovakia',
       'Egypt', 'Italy', 'Netherlands', 'Germany', 'Cuba', 'Hungary',
       'Spain', 'Switzerland', 'Mexico', 'England', 'Sweden', 'Scotland',
       'South Korea', 'Colombia', 'Russia', 'Bulgaria', 'North Korea',
       'Portugal', 'Israel', 'Morocco', 'El Salvador', 'DR Congo',
       'German DR', 'Poland', 'Yugoslavia', 'Haiti', 'Australia', 'Iran',
       'Algeria', 'Honduras', 'Canada', 'Northern Ireland', 'Denmark',
       'Iraq', 'United Arab Emirates', 'United States', 'Costa Rica',
       'Cameroon', 'Republic of Ireland', 'Norway', 'Nigeria', 'Romania',
       'Saudi Arabia', 'Greece', 'Jamaica', 'Serbia', 'South Africa',
       'Japan', 'Croatia', 'China PR', 'Tunisia', 'Senegal', 'Slovenia',
       'Ecuador', 'Turkey', 'Trinidad and Tobago', 'Angola',
       'Czech Republic', 'Togo', 'Ivory Coast', 'Ghana', 'Ukraine',
       'New Zealand

- Agrupando o numero de gols por time e por ano.

In [251]:
score_fifa.groupby(['team','year']).score.sum()

team        year
Algeria     1982     5
            1986     1
            2010     0
            2014     7
Angola      2006     1
                    ..
Yugoslavia  1958     7
            1962    10
            1974    12
            1982     2
            1990     8
Name: score, Length: 489, dtype: int64

- Fazendo a mesma coisa que fiz acima, porém transformando em um DataFrame para visualizar melhor.

In [252]:
pd.DataFrame(score_fifa.groupby(['team','year']).score.sum())

score
team       year       
Algeria    1982      5
           1986      1
           2010      0
           2014      7
Angola     2006      1
...                ...
Yugoslavia 1958      7
           1962     10
           1974     12
           1982      2
           1990      8

[489 rows x 1 columns]

- Pivotando os dados da planilha para ficar da meneira correta de visualizar.

In [253]:
score_fifa.groupby(['team','year']).score.sum().unstack()

year                  1930  1934  1938  1950  1954  1958  1962  1966  1970  \
team                                                                         
Algeria                NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Angola                 NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Argentina             18.0   2.0   NaN   NaN   NaN   5.0   2.0   4.0   NaN   
Australia              NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Austria                NaN   7.0   NaN   NaN  17.0   2.0   NaN   NaN   NaN   
...                    ...   ...   ...   ...   ...   ...   ...   ...   ...   
United Arab Emirates   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
United States          7.0   1.0   NaN   4.0   NaN   NaN   NaN   NaN   NaN   
Uruguay               15.0   NaN   NaN  15.0  16.0   NaN   4.0   2.0   4.0   
Wales                  NaN   NaN   NaN   NaN   NaN   4.0   NaN   NaN   NaN   
Yugoslavia             7.0   NaN   NaN   7.0   2.0   7.0  10.0   NaN   NaN   

year                  1974  ...  1986  1990  1994  1998  2002  2006  2010  \
team                        ...                                             
Algeria                NaN  ...   1.0   NaN   NaN   NaN   NaN   NaN   0.0   
Angola                 NaN  ...   NaN   NaN   NaN   NaN   NaN   1.0   NaN   
Argentina              9.0  ...  14.0   5.0   8.0  10.0   2.0  11.0  10.0   
Australia              0.0  ...   NaN   NaN   NaN   NaN   NaN   5.0   3.0   
Austria                NaN  ...   NaN   2.0   NaN   3.0   NaN   NaN   NaN   
...                    ...  ...   ...   ...   ...   ...   ...   ...   ...   
United Arab Emirates   NaN  ...   NaN   2.0   NaN   NaN   NaN   NaN   NaN   
United States          NaN  ...   NaN   2.0   3.0   1.0   7.0   2.0   5.0   
Uruguay                1.0  ...   2.0   2.0   NaN   NaN   4.0   NaN  11.0   
Wales                  NaN  ...   NaN   NaN   NaN   NaN   NaN   NaN   NaN   
Yugoslavia            12.0  ...   NaN   8.0   NaN   NaN   NaN   NaN   NaN   

year                  2014  2018  2022  
team                                    
Algeria                7.0   NaN   NaN  
Angola                 NaN   NaN   NaN  
Argentina              8.0   6.0   7.0  
Australia              3.0   2.0   4.0  
Austria                NaN   NaN   NaN  
...                    ...   ...   ...  
United Arab Emirates   NaN   NaN   NaN  
United States          5.0   NaN   3.0  
Uruguay                4.0   7.0   2.0  
Wales                  NaN   NaN   1.0  
Yugoslavia             NaN   NaN   NaN  

[82 rows x 22 columns]

- Como pivotamos a planilha, apareceu uma quantidade enorme de valores NaN (Not a Number), pois os países que não participaram da copa no ano ou não fizeram gol vai aparecere que não existe dado, já na forma que estávamos visualizando antes os anos que as seleções não marcavam gols ou não participaram da copa não aparecem.

- Usamos um 'fillna' para resolver isso.

In [254]:
score_fifa.groupby(['team','year']).score.sum().unstack().fillna(0)

year                  1930  1934  1938  1950  1954  1958  1962  1966  1970  \
team                                                                         
Algeria                0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
Angola                 0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
Argentina             18.0   2.0   0.0   0.0   0.0   5.0   2.0   4.0   0.0   
Australia              0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
Austria                0.0   7.0   0.0   0.0  17.0   2.0   0.0   0.0   0.0   
...                    ...   ...   ...   ...   ...   ...   ...   ...   ...   
United Arab Emirates   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
United States          7.0   1.0   0.0   4.0   0.0   0.0   0.0   0.0   0.0   
Uruguay               15.0   0.0   0.0  15.0  16.0   0.0   4.0   2.0   4.0   
Wales                  0.0   0.0   0.0   0.0   0.0   4.0   0.0   0.0   0.0   
Yugoslavia             7.0   0.0   0.0   7.0   2.0   7.0  10.0   0.0   0.0   

year                  1974  ...  1986  1990  1994  1998  2002  2006  2010  \
team                        ...                                             
Algeria                0.0  ...   1.0   0.0   0.0   0.0   0.0   0.0   0.0   
Angola                 0.0  ...   0.0   0.0   0.0   0.0   0.0   1.0   0.0   
Argentina              9.0  ...  14.0   5.0   8.0  10.0   2.0  11.0  10.0   
Australia              0.0  ...   0.0   0.0   0.0   0.0   0.0   5.0   3.0   
Austria                0.0  ...   0.0   2.0   0.0   3.0   0.0   0.0   0.0   
...                    ...  ...   ...   ...   ...   ...   ...   ...   ...   
United Arab Emirates   0.0  ...   0.0   2.0   0.0   0.0   0.0   0.0   0.0   
United States          0.0  ...   0.0   2.0   3.0   1.0   7.0   2.0   5.0   
Uruguay                1.0  ...   2.0   2.0   0.0   0.0   4.0   0.0  11.0   
Wales                  0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
Yugoslavia            12.0  ...   0.0   8.0   0.0   0.0   0.0   0.0   0.0   

year                  2014  2018  2022  
team                                    
Algeria                7.0   0.0   0.0  
Angola                 0.0   0.0   0.0  
Argentina              8.0   6.0   7.0  
Australia              3.0   2.0   4.0  
Austria                0.0   0.0   0.0  
...                    ...   ...   ...  
United Arab Emirates   0.0   0.0   0.0  
United States          5.0   0.0   3.0  
Uruguay                4.0   7.0   2.0  
Wales                  0.0   0.0   1.0  
Yugoslavia             0.0   0.0   0.0  

[82 rows x 22 columns]

In [255]:
score_fifa.groupby(['team','year']).score.sum().unstack().fillna(0).cumsum(axis=1).reset_index()

year                  team  1930  1934  1938  1950  1954  1958  1962  1966  \
0                  Algeria   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1                   Angola   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2                Argentina  18.0  20.0  20.0  20.0  20.0  25.0  27.0  31.0   
3                Australia   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4                  Austria   0.0   7.0   7.0   7.0  24.0  26.0  26.0  26.0   
..                     ...   ...   ...   ...   ...   ...   ...   ...   ...   
77    United Arab Emirates   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
78           United States   7.0   8.0   8.0  12.0  12.0  12.0  12.0  12.0   
79                 Uruguay  15.0  15.0  15.0  30.0  46.0  46.0  50.0  52.0   
80                   Wales   0.0   0.0   0.0   0.0   0.0   4.0   4.0   4.0   
81              Yugoslavia   7.0   7.0   7.0  14.0  16.0  23.0  33.0  33.0   

year  1970  ...  1986  1990  1994   1998   2002   2006   2010   2014   2018  \
0      0.0  ...   6.0   6.0   6.0    6.0    6.0    6.0    6.0   13.0   13.0   
1      0.0  ...   0.0   0.0   0.0    0.0    0.0    1.0    1.0    1.0    1.0   
2     31.0  ...  77.0  82.0  90.0  100.0  102.0  113.0  123.0  131.0  137.0   
3      0.0  ...   0.0   0.0   0.0    0.0    0.0    5.0    8.0   11.0   13.0   
4     26.0  ...  38.0  40.0  40.0   43.0   43.0   43.0   43.0   43.0   43.0   
..     ...  ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
77     0.0  ...   0.0   2.0   2.0    2.0    2.0    2.0    2.0    2.0    2.0   
78    12.0  ...  12.0  14.0  17.0   18.0   25.0   27.0   32.0   37.0   37.0   
79    56.0  ...  59.0  61.0  61.0   61.0   65.0   65.0   76.0   80.0   87.0   
80     4.0  ...   4.0   4.0   4.0    4.0    4.0    4.0    4.0    4.0    4.0   
81    33.0  ...  47.0  55.0  55.0   55.0   55.0   55.0   55.0   55.0   55.0   

year   2022  
0      13.0  
1       1.0  
2     144.0  
3      17.0  
4      43.0  
..      ...  
77      2.0  
78     40.0  
79     89.0  
80      5.0  
81     55.0  

[82 rows x 23 columns]

In [256]:
score_fifa_year = score_fifa.groupby(['team','year']).score.sum().unstack().fillna(0).cumsum(axis=1).reset_index()
score_fifa_year

year                  team  1930  1934  1938  1950  1954  1958  1962  1966  \
0                  Algeria   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1                   Angola   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2                Argentina  18.0  20.0  20.0  20.0  20.0  25.0  27.0  31.0   
3                Australia   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4                  Austria   0.0   7.0   7.0   7.0  24.0  26.0  26.0  26.0   
..                     ...   ...   ...   ...   ...   ...   ...   ...   ...   
77    United Arab Emirates   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
78           United States   7.0   8.0   8.0  12.0  12.0  12.0  12.0  12.0   
79                 Uruguay  15.0  15.0  15.0  30.0  46.0  46.0  50.0  52.0   
80                   Wales   0.0   0.0   0.0   0.0   0.0   4.0   4.0   4.0   
81              Yugoslavia   7.0   7.0   7.0  14.0  16.0  23.0  33.0  33.0   

year  1970  ...  1986  1990  1994   1998   2002   2006   2010   2014   2018  \
0      0.0  ...   6.0   6.0   6.0    6.0    6.0    6.0    6.0   13.0   13.0   
1      0.0  ...   0.0   0.0   0.0    0.0    0.0    1.0    1.0    1.0    1.0   
2     31.0  ...  77.0  82.0  90.0  100.0  102.0  113.0  123.0  131.0  137.0   
3      0.0  ...   0.0   0.0   0.0    0.0    0.0    5.0    8.0   11.0   13.0   
4     26.0  ...  38.0  40.0  40.0   43.0   43.0   43.0   43.0   43.0   43.0   
..     ...  ...   ...   ...   ...    ...    ...    ...    ...    ...    ...   
77     0.0  ...   0.0   2.0   2.0    2.0    2.0    2.0    2.0    2.0    2.0   
78    12.0  ...  12.0  14.0  17.0   18.0   25.0   27.0   32.0   37.0   37.0   
79    56.0  ...  59.0  61.0  61.0   61.0   65.0   65.0   76.0   80.0   87.0   
80     4.0  ...   4.0   4.0   4.0    4.0    4.0    4.0    4.0    4.0    4.0   
81    33.0  ...  47.0  55.0  55.0   55.0   55.0   55.0   55.0   55.0   55.0   

year   2022  
0      13.0  
1       1.0  
2     144.0  
3      17.0  
4      43.0  
..      ...  
77      2.0  
78     40.0  
79     89.0  
80      5.0  
81     55.0  

[82 rows x 23 columns]

- Importando os dados das bandeiras, confederações e siglas dos países.

In [257]:
flags = pd.read_csv('countries-fifa-flags.csv')
flags

code            country confederation  \
0    AFG        Afghanistan           AFC   
1    AIA           Anguilla       CONCACF   
2    ALB            Albania          UEFA   
3    ALG            Algeria           CAF   
4    ASA     American Samoa           OFC   
..   ...                ...           ...   
206  ISV  US Virgin Islands       CONCACF   
207  WAL              Wales          UEFA   
208  YEM              Yemen           AFC   
209  ZAM             Zambia           CAF   
210  ZIM           Zimbabwe           CAF   

                                                   url  
0          https://countryflagsapi.com/png/afghanistan  
1             https://countryflagsapi.com/png/anguilla  
2              https://countryflagsapi.com/png/albania  
3              https://countryflagsapi.com/png/algeria  
4     https://countryflagsapi.com/png/american%20samoa  
..                                                 ...  
206  https://countryflagsapi.com/png/us%20virgin%20...  
207              https://countryflagsapi.com/png/wales  
208              https://countryflagsapi.com/png/yemen  
209             https://countryflagsapi.com/png/zambia  
210           https://countryflagsapi.com/png/zimbabwe  

[211 rows x 4 columns]

- Unindo dois DataFrames lateralmente como 'merge'.

In [269]:
pd.merge(right=flags,left=score_fifa_year,left_on='team',right_on='country')


team  1930  1934  1938  1950  1954  1958  1962  1966  \
0                Algeria   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1                 Angola   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2              Argentina  18.0  20.0  20.0  20.0  20.0  25.0  27.0  31.0   
3              Australia   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4                Austria   0.0   7.0   7.0   7.0  24.0  26.0  26.0  26.0   
..                   ...   ...   ...   ...   ...   ...   ...   ...   ...   
65                Turkey   0.0   0.0   0.0   0.0  10.0  10.0  10.0  10.0   
66               Ukraine   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
67  United Arab Emirates   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
68               Uruguay  15.0  15.0  15.0  30.0  46.0  46.0  50.0  52.0   
69                 Wales   0.0   0.0   0.0   0.0   0.0   4.0   4.0   4.0   

    1970  ...   2002   2006   2010   2014   2018   2022  code  \
0    0.0  ...    6.0    6.0    6.0   13.0   13.0   13.0   ALG   
1    0.0  ...    0.0    1.0    1.0    1.0    1.0    1.0   ANT   
2   31.0  ...  102.0  113.0  123.0  131.0  137.0  144.0   ARG   
3    0.0  ...    0.0    5.0    8.0   11.0   13.0   17.0   AUS   
4   26.0  ...   43.0   43.0   43.0   43.0   43.0   43.0   AUT   
..   ...  ...    ...    ...    ...    ...    ...    ...   ...   
65  10.0  ...   20.0   20.0   20.0   20.0   20.0   20.0   TUR   
66   0.0  ...    0.0    5.0    5.0    5.0    5.0    5.0   UKR   
67   0.0  ...    2.0    2.0    2.0    2.0    2.0    2.0   UAE   
68  56.0  ...   65.0   65.0   76.0   80.0   87.0   89.0   URU   
69   4.0  ...    4.0    4.0    4.0    4.0    4.0    5.0   WAL   

                 country  confederation  \
0                Algeria            CAF   
1                 Angola            CAF   
2              Argentina       CONMEBOL   
3              Australia            OFC   
4                Austria           UEFA   
..                   ...            ...   
65                Turkey           UEFA   
66               Ukraine           UEFA   
67  United Arab Emirates            AFC   
68               Uruguay       CONMEBOL   
69                 Wales           UEFA   

                                                  url  
0             https://countryflagsapi.com/png/algeria  
1              https://countryflagsapi.com/png/angola  
2           https://countryflagsapi.com/png/argentina  
3           https://countryflagsapi.com/png/australia  
4             https://countryflagsapi.com/png/austria  
..                                                ...  
65             https://countryflagsapi.com/png/turkey  
66            https://countryflagsapi.com/png/ukraine  
67  https://countryflagsapi.com/png/united%20arab%...  
68            https://countryflagsapi.com/png/uruguay  
69              https://countryflagsapi.com/png/wales  

[70 rows x 27 columns]

- Analisamos qu nem todas as linhas foram para o novo DF.

Vamos analisar quais linhas não foram adicionadas com a função 'isin'.

In [264]:
score_fifa_year[~score_fifa_year['team'].isin(flags.country)]

year                 team  1930  1934  1938  1950  1954  1958  1962  1966  \
13               China PR   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
19         Czechoslovakia   0.0   9.0  14.0  14.0  14.0  23.0  30.0  30.0   
20               DR Congo   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
27              German DR   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
36                   Iran   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
40            Ivory Coast   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
49            North Korea   0.0   0.0   0.0   0.0   0.0   0.0   0.0   5.0   
58    Republic of Ireland   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
60                 Russia   0.0   0.0   0.0   0.0   0.0   5.0  14.0  24.0   
68            South Korea   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
78          United States   7.0   8.0   8.0  12.0  12.0  12.0  12.0  12.0   
81             Yugoslavia   7.0   7.0   7.0  14.0  16.0  23.0  33.0  33.0   

year  1970  ...  1986  1990  1994  1998  2002  2006  2010  2014  2018  2022  
13     0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
19    32.0  ...  34.0  44.0  44.0  44.0  44.0  44.0  44.0  44.0  44.0  44.0  
20     0.0  ...   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0  
27     0.0  ...   5.0   5.0   5.0   5.0   5.0   5.0   5.0   5.0   5.0   5.0  
36     0.0  ...   2.0   2.0   2.0   4.0   4.0   6.0   6.0   7.0   9.0  13.0  
40     0.0  ...   0.0   0.0   0.0   0.0   0.0   5.0   9.0  13.0  13.0  13.0  
49     5.0  ...   5.0   5.0   5.0   5.0   5.0   5.0   6.0   6.0   6.0   6.0  
58     0.0  ...   0.0   2.0   4.0   4.0  10.0  10.0  10.0  10.0  10.0  10.0  
60    30.0  ...  49.0  53.0  60.0  60.0  64.0  64.0  64.0  66.0  77.0  77.0  
68     0.0  ...   4.0   5.0   9.0  11.0  19.0  22.0  28.0  31.0  34.0  39.0  
78    12.0  ...  12.0  14.0  17.0  18.0  25.0  27.0  32.0  37.0  37.0  40.0  
81    33.0  ...  47.0  55.0  55.0  55.0  55.0  55.0  55.0  55.0  55.0  55.0  

[12 rows x 23 columns]

- Para podermos colocar as colunas ignorando o erro gerado pela diferença entre os nomes de flags.country e score_fifa_year.team, vamos usar o parâmetro 'how' e indicar qual DataFrame queremos usar como fixo.

In [270]:
dados = pd.merge(right=flags,left=score_fifa_year,left_on='team',right_on='country', how='left')
dados

team  1930  1934  1938  1950  1954  1958  1962  1966  \
0                Algeria   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
1                 Angola   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
2              Argentina  18.0  20.0  20.0  20.0  20.0  25.0  27.0  31.0   
3              Australia   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
4                Austria   0.0   7.0   7.0   7.0  24.0  26.0  26.0  26.0   
..                   ...   ...   ...   ...   ...   ...   ...   ...   ...   
77  United Arab Emirates   0.0   0.0   0.0   0.0   0.0   0.0   0.0   0.0   
78         United States   7.0   8.0   8.0  12.0  12.0  12.0  12.0  12.0   
79               Uruguay  15.0  15.0  15.0  30.0  46.0  46.0  50.0  52.0   
80                 Wales   0.0   0.0   0.0   0.0   0.0   4.0   4.0   4.0   
81            Yugoslavia   7.0   7.0   7.0  14.0  16.0  23.0  33.0  33.0   

    1970  ...   2002   2006   2010   2014   2018   2022  code  \
0    0.0  ...    6.0    6.0    6.0   13.0   13.0   13.0   ALG   
1    0.0  ...    0.0    1.0    1.0    1.0    1.0    1.0   ANT   
2   31.0  ...  102.0  113.0  123.0  131.0  137.0  144.0   ARG   
3    0.0  ...    0.0    5.0    8.0   11.0   13.0   17.0   AUS   
4   26.0  ...   43.0   43.0   43.0   43.0   43.0   43.0   AUT   
..   ...  ...    ...    ...    ...    ...    ...    ...   ...   
77   0.0  ...    2.0    2.0    2.0    2.0    2.0    2.0   UAE   
78  12.0  ...   25.0   27.0   32.0   37.0   37.0   40.0   NaN   
79  56.0  ...   65.0   65.0   76.0   80.0   87.0   89.0   URU   
80   4.0  ...    4.0    4.0    4.0    4.0    4.0    5.0   WAL   
81  33.0  ...   55.0   55.0   55.0   55.0   55.0   55.0   NaN   

                 country  confederation  \
0                Algeria            CAF   
1                 Angola            CAF   
2              Argentina       CONMEBOL   
3              Australia            OFC   
4                Austria           UEFA   
..                   ...            ...   
77  United Arab Emirates            AFC   
78                   NaN            NaN   
79               Uruguay       CONMEBOL   
80                 Wales           UEFA   
81                   NaN            NaN   

                                                  url  
0             https://countryflagsapi.com/png/algeria  
1              https://countryflagsapi.com/png/angola  
2           https://countryflagsapi.com/png/argentina  
3           https://countryflagsapi.com/png/australia  
4             https://countryflagsapi.com/png/austria  
..                                                ...  
77  https://countryflagsapi.com/png/united%20arab%...  
78                                                NaN  
79            https://countryflagsapi.com/png/uruguay  
80              https://countryflagsapi.com/png/wales  
81                                                NaN  

[82 rows x 27 columns]

- Vamos salvar o nosso DF filtrado, organizado e ajustado para mandar par ao Flourish

In [285]:
dados.to_csv('fifa_data_viz.csv',index=False)